### ISSUE TITLE
Fix S4159: Rule should handle generic interfaces correctly

### ISSUE THREAD

**Issue by Evangelink** [SonarSource/sonar-csharp#1296]
### Description

Rule raises FP when implementing generic interface.

### Repro steps

```
interface ISomething<T> { }
public class BaseThing { }

[Export(typeof(ISomething<BaseThing>))]
public class BaseSomethingImplementation : ISomething<BaseThing> // Noncompliant
{
}

[Export(typeof(ISomething<>))]
public class Something<T> : ISomething<T>
{
}
```

### Expected behavior

Rule doesn't raise an issue.

### Actual behavior

Issue is raised.

### Related information

See [SonarLint Google Group](https://groups.google.com/d/msgid/sonarlint/d4fb879a-a561-4b3c-8ba9-cb139b27cc8b%40googlegroups.com?utm_medium=email&utm_source=footer)
